File: src/main/java/com/ml/atomic/mathgame/ArithmeticGenerator.java
package com.ml.atomic.mathgame;

import com.google.common.collect.Lists;
import com.google.common.collect.Range;
import com.google.common.collect.Sets;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.extern.java.Log;
import org.apache.commons.lang3.time.StopWatch;

import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

import static org.apache.commons.lang3.RandomUtils.nextInt;

/**
 * 算术题生成器
 */
@Log
class ArithmeticGenerator {

    List<Result> execute(Configuration configuration) {

        StopWatch sw = new StopWatch();
        sw.start();
        log.info("使用配置".concat(configuration.toString()).concat("开始生成算术题..."));

        if (configuration.subItems.isEmpty())
            throw new IllegalArgumentException("至少提供一个生成序列");

        // 使用set避免重复
        int totalCount = configuration.getTotalCount();
        Set<Result> result = Sets.newHashSetWithExpectedSize(totalCount);

        // 每一条子配置生成count条
        configuration.subItems.forEach(config -> {
            AtomicInteger subSeqCount = new AtomicInteger();
            do {
                doGenerateItem(configuration.getFirstNumRange(), configuration.getResultRange(), config).ifPresent(resultItem -> {
                    result.add(resultItem);
                    subSeqCount.getAndIncrement();
                });
            } while (subSeqCount.get() < config.getCount());
        });

        sw.stop();
        log.info("生成[" + totalCount + "]条算术题完毕, 总耗时[ " + sw.getTime() + " ]ms");
        return Lists.newArrayList(result);
    }

    /**
     * 生成一条完整序列的算式
     */
    private static Optional<Result> doGenerateItem(Range<Integer> firstNum, Range<Integer> resultRange, ConfigurationPart subSeq) {

        List<CalcStep> steps = subSeq.steps;
        int first = nextInt(firstNum.lowerEndpoint(), firstNum.upperEndpoint());
        do {
            List<CalcStep.Result> resultHolder = Lists.newArrayListWithCapacity(steps.size());
            for (int i = 0; i < steps.size(); i++) {
                CalcStep step = steps.get(i);
                int lastVal = i == 0 ? first : resultHolder.get(i - 1).result;
                Optional<CalcStep.Result> stepResult = step.apply(lastVal);

                if (stepResult.isPresent()) resultHolder.add(stepResult.get());
                else return Optional.empty();
            }
            int finalResult = resultHolder.get(resultHolder.size() - 1).result;

            if (resultRange.contains(finalResult)) { // 落到了合理的区间内
                String other = resultHolder.stream().map(i -> String.format("%s %s", i.operator.label, i.val)).collect(Collectors.joining(" ", " ", " "));

                String blank = String.format("%s%s=", first, other);
                String withAnswer = String.format("%s%s= %s", first, other, finalResult);

                return Optional.of(new Result(blank, withAnswer));

            }
        } while (true);

    }

    @AllArgsConstructor
    @Getter
    static class Result {
        String blank;
        String withAnswer;
    }

}

File: arithmetic_generator.py
from calc_step import Result
import time
from random import randint
from typing import List, Optional

import portion

from configuration import Configuration, ConfigurationPart


class ArithmeticGenerator(object):
    """ 算术题生成器 """

    def execute(self, config: Configuration):
        start_time = time.time()

        if len(config.sub_items) == 0:
            raise ValueError("至少提供一个生成序列")

        total_count = config.total_count()
        print("使用配置config = (", config, ")，开始生成算术题...")

        result = []
        for item in config.sub_items:
            sub_seq_count = 0
            while sub_seq_count < item.count:
                result_item = self.do_generate_item(config.first_num_range, config.result_range, item)
                if result_item is not None:
                    result.append(result_item)
                    sub_seq_count += 1

        cost = 1000 * (time.time() - start_time)
        print("生成[", total_count, "]条算术题完毕, 共耗时 = ",
              cost, "ms")
        return result

    def do_generate_item(self, first_num: portion, result_range: portion, sub_seq: ConfigurationPart):
        """
        生成一条完整序列的算式
        """
        first = randint(first_num.lower, first_num.upper)
        while True:
            result_holder: List[Result] = []
            for i, step in enumerate(sub_seq.steps):
                last_val = first if i == 0 else result_holder[-1]["result"]
                step_result = step.apply(last_val)
                if step_result is not None:
                    result_holder.append(step_result)
                else:
                    return

            # 序列生成结束后，最后计算结果是最后一个元素
            final_result = result_holder[-1]["result"]
            if final_result in result_range:
                # 组合后续序列的值
                other = " ".join(map(lambda i: "{} {}".format(i["operator"].value, i["val"]), result_holder))
                blank = "{} {} =".format(first, other)
                with_answer = "{} {} = {}".format(first, other, final_result)
                # print(with_answer)

                return (blank, with_answer)
